File: ios/Mailrs/Accounts/IMAPLine.swift
import Foundation

/// Reading what an IMAP server says.
///
/// Split from the socket so it can be tested without one: every
/// mistake worth making here is in the parsing, and a test that needs
/// a server is a test nobody runs.
enum IMAP {
    /// One untagged line, as far as this client cares about it.
    enum Untagged: Equatable {
        /// `* LIST (\HasNoChildren \Sent) "/" "[Gmail]/Sent Mail"`
        case list(name: String, attributes: [String])
        /// `* 42 EXISTS`
        case exists(Int)
        /// `* OK [UIDVALIDITY 1234] ...`
        case uidValidity(UInt32)
        /// `* OK [UIDNEXT 4391] ...`
        case uidNext(UInt32)
        /// What the server says it can do.
        ///
        /// Announced in two places — the greeting and a `CAPABILITY`
        /// reply — and a client that reads only the second asks a
        /// question it already has the answer to.
        case capabilities([String])
        /// Anything this client has no use for.
        case other
    }

    /// How a tagged line ended.
    enum Completion: Equatable {
        case ok(String)
        case no(String)
        case bad(String)
    }

    /// The tagged reply for `tag`, or nil if this is not one.
    ///
    /// A tag is compared with a trailing space: `a1` must not match
    /// `a10`, and a server is free to interleave `a10`'s reply while
    /// `a1` is outstanding.
    static func completion(of line: String, tag: String) -> Completion? {
        guard line.hasPrefix(tag + " ") else { return nil }
        let rest = String(line.dropFirst(tag.count + 1))
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let word = rest.split(separator: " ", maxSplits: 1).first.map(String.init) ?? ""
        let detail = rest.dropFirst(word.count).trimmingCharacters(in: .whitespaces)
        switch word.uppercased() {
        case "OK": return .ok(detail)
        case "NO": return .no(detail)
        case "BAD": return .bad(detail)
        default: return nil
        }
    }

    /// What an untagged line says, as far as this client uses it.
    static func untagged(_ line: String) -> Untagged? {
        guard line.hasPrefix("* ") else { return nil }
        let body = String(line.dropFirst(2)).trimmingCharacters(in: .whitespacesAndNewlines)

        if body.uppercased().hasPrefix("LIST ") {
            return parseList(body)
        }
        if body.uppercased().hasPrefix("CAPABILITY") {
            return .capabilities(body.split(separator: " ").dropFirst().map(String.init))
        }
        // Also announced inside the greeting's response code, which is
        // where a server that offers no `CAPABILITY` at all says it.
        if let open = body.range(of: "[CAPABILITY ", options: .caseInsensitive),
            let close = body[open.upperBound...].firstIndex(of: "]")
        {
            return .capabilities(
                body[open.upperBound..<close].split(separator: " ").map(String.init))
        }
        // `* 42 EXISTS` — a count, then the word.
        let parts = body.split(separator: " ", maxSplits: 1)
        if parts.count == 2, let n = Int(parts[0]),
           parts[1].uppercased().hasPrefix("EXISTS") {
            return .exists(n)
        }
        if let v = bracketed(body, "UIDVALIDITY") { return .uidValidity(v) }
        if let v = bracketed(body, "UIDNEXT") { return .uidNext(v) }
        return .other
    }

    /// `OK [UIDVALIDITY 1234] Ready` → 1234.
    ///
    /// The value is inside square brackets in a response code, and the
    /// text after it is free-form: a server may say anything at all,
    /// including something that looks like another number.
    private static func bracketed(_ body: String, _ key: String) -> UInt32? {
        guard let open = body.range(of: "[\(key) ", options: .caseInsensitive),
              let close = body.range(of: "]", range: open.upperBound..<body.endIndex)
        else { return nil }
        return UInt32(body[open.upperBound..<close.lowerBound]
            .trimmingCharacters(in: .whitespaces))
    }

    /// `LIST (\HasNoChildren \Sent) "/" "[Gmail]/Sent Mail"`
    ///
    /// The name is last and may be quoted, may contain spaces, and may
    /// contain the delimiter — which is why it is taken from the end
    /// rather than by splitting on spaces.
    private static func parseList(_ body: String) -> Untagged? {
        guard let open = body.firstIndex(of: "("),
              let close = body.firstIndex(of: ")"), open < close
        else { return nil }
        let attributes = body[body.index(after: open)..<close]
            .split(separator: " ")
            .map(String.init)
        let tail = body[body.index(after: close)...]
            .trimmingCharacters(in: .whitespaces)
        // delimiter then name; the delimiter is quoted or NIL
        guard let name = lastQuotedOrWord(tail) else { return nil }
        return .list(name: name, attributes: attributes)
    }

    /// The last field of a line, unquoted.
    ///
    /// `"/" "[Gmail]/Sent Mail"` → `[Gmail]/Sent Mail`. Taken from the
    /// end because a mailbox name may hold spaces and the delimiter.
    static func lastQuotedOrWord(_ s: String) -> String? {
        let t = s.trimmingCharacters(in: .whitespaces)
        guard !t.isEmpty else { return nil }
        if t.hasSuffix("\"") {
            // Walk **forwards**, tracking escapes, and remember where
            // the last unescaped quote opened. Walking backwards and
            // stopping at the first unescaped quote finds the closing
            // one of an empty pair when the name itself ends in `\"`.
            let chars = Array(t)
            var opens: [Int] = []
            var i = 0
            while i < chars.count {
                if chars[i] == "\\" {
                    i += 2
                    continue
                }
                if chars[i] == "\"" { opens.append(i) }
                i += 1
            }
            guard opens.count >= 2 else { return nil }
            let open = opens[opens.count - 2]
            let close = opens[opens.count - 1]
            return String(chars[(open + 1)..<close])
                .replacingOccurrences(of: "\\\"", with: "\"")
                .replacingOccurrences(of: "\\\\", with: "\\")
        }
        return t.split(separator: " ").last.map(String.init)
    }

    /// What a `FETCH` line announced, when it announced a literal.
    ///
    /// `* 12 FETCH (UID 4390 FLAGS (\Seen) BODY[] {2048}` — the
    /// braces give the byte count, and **the bytes that follow are
    /// read by that count rather than scanned for a terminator**. A
    /// message body contains every byte sequence a terminator could be
    /// made of, so scanning truncates mail at whatever looks like the
    /// end.
    struct Announced: Equatable {
        let uid: UInt32?
        let seen: Bool
        /// How many bytes follow, when the line ends in a literal.
        let literalBytes: Int?
    }

    /// Read a `FETCH` line.
    static func fetchLine(_ line: String) -> Announced? {
        guard line.hasPrefix("* "), line.uppercased().contains(" FETCH ") else { return nil }
        let uid = number(after: "UID ", in: line).map { UInt32($0) }
        // `\Seen` inside the FLAGS list. Matched with the backslash so
        // a folder called "Seen" in the same line cannot set it.
        let seen = line.uppercased().contains("\\SEEN")
        var literal: Int?
        if let open = line.lastIndex(of: "{"), let close = line.lastIndex(of: "}"),
           open < close {
            literal = Int(line[line.index(after: open)..<close])
        }
        return Announced(uid: uid, seen: seen, literalBytes: literal)
    }

    /// The number after a keyword, or nil.
    private static func number(after keyword: String, in line: String) -> Int? {
        guard let r = line.range(of: keyword, options: .caseInsensitive) else { return nil }
        let digits = line[r.upperBound...].prefix { $0.isNumber }
        return digits.isEmpty ? nil : Int(digits)
    }

    /// Quote a mailbox name or a password for the wire.
    ///
    /// Generated app passwords contain `"` and `\` often enough that an
    /// unquoted argument turns one into a syntax error — and the person
    /// is told their password is wrong when it is right.
    static func quoted(_ s: String) -> String {
        let escaped = s
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "\"", with: "\\\"")
        return "\"\(escaped)\""
    }

    /// Whether a refusal means the credential is wrong rather than the
    /// server being unhappy about something else.
    ///
    /// Worth telling apart: one is a button to press, the other is
    /// waiting. RFC 5530 gives the code; older servers only say it in
    /// the text.
    static func isAuthenticationFailure(_ detail: String) -> Bool {
        let d = detail.uppercased()
        return d.contains("[AUTHENTICATIONFAILED]")
            || d.contains("AUTHENTICATIONFAILED")
            || d.contains("INVALID CREDENTIALS")
            || d.contains("LOGIN FAILED")
            || d.contains("[AUTHORIZATIONFAILED]")
    }
}

File: android/app/src/main/kotlin/jp/golia/mailrs/accounts/ImapLine.kt
package jp.golia.mailrs.accounts

/**
 * Reading what an IMAP server says.
 *
 * Split from the socket so it can be tested without one: every mistake
 * worth making here is in the parsing, and a test that needs a server
 * is a test nobody runs. The same split, and the same cases, as the
 * iOS side — a client that disagrees with itself across platforms is
 * two clients.
 */
object Imap {
    /** One untagged line, as far as this client cares about it. */
    sealed interface Untagged {
        data class ListFolder(val name: String, val attributes: List<String>) : Untagged
        data class Exists(val count: Int) : Untagged
        data class UidValidity(val value: Long) : Untagged
        data class UidNext(val value: Long) : Untagged
        /**
         * What the server says it can do.
         *
         * Announced in two places — the greeting and a `CAPABILITY`
         * reply — and a client that reads only the second asks a
         * question it already has the answer to.
         */
        data class Capabilities(val names: List<String>) : Untagged

        data object Other : Untagged
    }

    /** How a tagged line ended. */
    sealed interface Completion {
        data class Ok(val detail: String) : Completion
        data class No(val detail: String) : Completion
        data class Bad(val detail: String) : Completion
    }

    /**
     * The tagged reply for [tag], or null if this is not one.
     *
     * The tag is matched with a trailing space: `a1` must not match
     * `a10`, and a server is free to interleave `a10`'s reply while
     * `a1` is outstanding.
     */
    fun completion(line: String, tag: String): Completion? {
        if (!line.startsWith("$tag ")) return null
        val rest = line.removePrefix("$tag ").trim()
        val word = rest.substringBefore(' ', rest)
        val detail = rest.removePrefix(word).trim()
        return when (word.uppercase()) {
            "OK" -> Completion.Ok(detail)
            "NO" -> Completion.No(detail)
            "BAD" -> Completion.Bad(detail)
            else -> null
        }
    }

    /** What an untagged line says, as far as this client uses it. */
    fun untagged(line: String): Untagged? {
        if (!line.startsWith("* ")) return null
        val body = line.removePrefix("* ").trim()

        if (body.uppercase().startsWith("LIST ")) return parseList(body)
        if (body.uppercase().startsWith("CAPABILITY")) {
            return Untagged.Capabilities(
                body.split(" ").drop(1).filter { it.isNotEmpty() },
            )
        }
        // Also announced inside the greeting's response code, which is
        // where a server that offers no `CAPABILITY` at all says it.
        val open = body.indexOf("[CAPABILITY ", ignoreCase = true)
        if (open >= 0) {
            val close = body.indexOf(']', open)
            if (close > open) {
                return Untagged.Capabilities(
                    body.substring(open + "[CAPABILITY ".length, close)
                        .split(" ").filter { it.isNotEmpty() },
                )
            }
        }

        val head = body.substringBefore(' ')
        val tail = body.removePrefix(head).trim()
        val n = head.toIntOrNull()
        if (n != null && tail.uppercase().startsWith("EXISTS")) return Untagged.Exists(n)

        bracketed(body, "UIDVALIDITY")?.let { return Untagged.UidValidity(it) }
        bracketed(body, "UIDNEXT")?.let { return Untagged.UidNext(it) }
        return Untagged.Other
    }

    /**
     * `OK [UIDVALIDITY 1234] Ready` -> 1234.
     *
     * The value is inside brackets in a response code, and the text
     * after it is free-form: a server may say anything at all,
     * including something that looks like another number.
     */
    private fun bracketed(body: String, key: String): Long? {
        val open = body.indexOf("[$key ", ignoreCase = true)
        if (open < 0) return null
        val from = open + key.length + 2
        val close = body.indexOf(']', from)
        if (close < 0) return null
        return body.substring(from, close).trim().toLongOrNull()
    }

    /**
     * `LIST (\HasNoChildren \Sent) "/" "[Gmail]/Sent Mail"`
     *
     * The name is last and may be quoted, may contain spaces, and may
     * contain the delimiter — which is why it is taken from the end
     * rather than by splitting on spaces.
     */
    private fun parseList(body: String): Untagged? {
        val open = body.indexOf('(')
        val close = body.indexOf(')')
        if (open < 0 || close < open) return null
        val attributes = body.substring(open + 1, close).split(' ').filter { it.isNotEmpty() }
        val name = lastQuotedOrWord(body.substring(close + 1).trim()) ?: return null
        return Untagged.ListFolder(name, attributes)
    }

    /**
     * The last field of a line, unquoted.
     *
     * Scanned **forwards** tracking escapes: walking back from the end
     * and stopping at the first unescaped quote finds the wrong one
     * when the name itself ends in `\"`.
     */
    fun lastQuotedOrWord(s: String): String? {
        val t = s.trim()
        if (t.isEmpty()) return null
        if (!t.endsWith('"')) return t.split(' ').lastOrNull()?.ifEmpty { null }
        val opens = mutableListOf<Int>()
        var i = 0
        while (i < t.length) {
            when {
                t[i] == '\\' -> i++
                t[i] == '"' -> opens += i
            }
            i++
        }
        if (opens.size < 2) return null
        return t.substring(opens[opens.size - 2] + 1, opens[opens.size - 1])
            .replace("\\\"", "\"")
            .replace("\\\\", "\\")
    }

    /**
     * What a `FETCH` line announced, when it announced a literal.
     *
     * `* 12 FETCH (UID 4390 FLAGS (\Seen) BODY[] {2048}` — the braces
     * give the byte count, and **the bytes that follow are read by
     * that count rather than scanned for a terminator**. A message
     * body contains every byte sequence a terminator could be made of,
     * so scanning truncates mail at whatever looks like the end.
     */
    data class Announced(val uid: Long?, val seen: Boolean, val literalBytes: Int?)

    /** Read a `FETCH` line. */
    fun fetchLine(line: String): Announced? {
        if (!line.startsWith("* ") || !line.uppercase().contains(" FETCH ")) return null
        val uid = numberAfter("UID ", line)
        // `\Seen` inside the FLAGS list. Matched with the backslash so
        // a folder called "Seen" in the same line cannot set it.
        val seen = line.uppercase().contains("\\SEEN")
        val open = line.lastIndexOf('{')
        val close = line.lastIndexOf('}')
        val literal = if (open in 0..<close) {
            line.substring(open + 1, close).toIntOrNull()
        } else {
            null
        }
        return Announced(uid, seen, literal)
    }

    /** The number after a keyword, or null. */
    private fun numberAfter(keyword: String, line: String): Long? {
        val at = line.indexOf(keyword, ignoreCase = true)
        if (at < 0) return null
        val digits = line.substring(at + keyword.length).takeWhile { it.isDigit() }
        return digits.toLongOrNull()
    }

    /**
     * Quote a mailbox name or a password for the wire.
     *
     * Generated app passwords contain `"` and `\` often enough that an
     * unquoted argument turns one into a syntax error — and the person
     * is told their password is wrong when it is right.
     */
    fun quoted(s: String): String =
        "\"" + s.replace("\\", "\\\\").replace("\"", "\\\"") + "\""

    /**
     * Whether a refusal means the credential is wrong rather than the
     * server being unhappy about something else.
     *
     * One is a button to press, the other is waiting.
     */
    fun isAuthenticationFailure(detail: String): Boolean {
        val d = detail.uppercase()
        return d.contains("AUTHENTICATIONFAILED") ||
            d.contains("INVALID CREDENTIALS") ||
            d.contains("LOGIN FAILED") ||
            d.contains("AUTHORIZATIONFAILED")
    }
}
